Extended flip() to more texture types and to S3TC compression

Vtec234 opened the pull request:
- Added flip() for texture2DArray, textureCube, textureCubeArray
- Added flip_s3tc() and flip_s3tc_block() for S3TC compressed images
- DXT1 compression with an alpha channel is NOT supported and causes an assertion failure. i'm happy to add it, but can't find any docs on how it is supposed to be laid out in memory

Groovounet commented:
I think a great improvement for this pull request would be to add your tests of this feature in test/core/core_flip.cpp

Beside this, this is great!

Vtec234 commented:
I added the tests for textureCube, textureCubeArray, RGB DXT1 and RGBA DXT3, DXT5.
Also made the flip_s3tc readable by using some structs and fixed some bugs that were in there.
RGBA DXT1 still doesn't work.

elect86 commented:
I always wondered how you can recognize from the header if it is a RGB or a RGBA DXT1

Armada651 commented on gli/core/flip.inl:
```diff
@@ -45,27 +46,214 @@ namespace detail
+		if(Format == FORMAT_RGBA_DXT1_UNORM || Format == FORMAT_RGBA_DXT1_SRGB)
+		{
+			// ?? can't find any spec for how a DXT1 texture with alpha is laid out
+			assert(false);
```
There is no difference in the layout between `FORMAT_RGB_DXT1` and `FORMAT_RGBA_DXT1`. The only difference is the interpretation of the data.

When rendering in the RGB format the alpha bit results in a black RGB color with an alpha value of 1. With RGBA it results in a black RGB color with an alpha value of 0. Both formats counter-intuitively support alpha bits, they're only sampled differently.

Armada651 commented:
@elect86 You can't recognize it from the header, because the RGB format does not exist for DirectX and is OpenGL-specific. See PR #71.

Armada651 commented:
@Groovounet This PR can be merged if you make the change I mentioned above before merging.

Vtec234 commented:
Fixed the DXT1 format flipping, this PR is now good to merge (some minor git conflicts need to be resolved).

stfx commented:
@Groovounet @Vtec234 any news regarding this nice PR?

Vtec234 commented:
@stfx Unfortunately there was no response from @Groovounet for quite a long time, so I closed it. Reopened it in #94 per your request.
